File: 0973-k-closest-points-to-origin/0973-k-closest-points-to-origin.cpp
// class Solution {
// public:
//     vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
//         vector<vector<int>> res(k);
//         vector<pair<int,pair<int,int>>> v;
//         int dist;
//         for(auto it:points)
//         {
//             dist=it[0]*it[0]+it[1]*it[1];
//             v.push_back({dist,{it[0],it[1]}});
//         }
//         sort(v.begin(),v.end());
//         for(int i=0;i<k;++i)
//         {
//             res[i].push_back(v[i].second.first);
//             res[i].push_back(v[i].second.second);
//         }
//         return res;
//     }
// };

class Solution {
public:
    vector<vector<int>> kClosest(vector<vector<int>>& points, int K) {
        priority_queue<vector<int>, vector<vector<int>>, compare> pq(points.begin(), points.end());
        vector<vector<int>> ans;
        for (int i = 0; i < K; i++) {
            ans.push_back(pq.top());
            pq.pop();
        }
        return ans;
    }
private:
    struct compare {
        bool operator()(vector<int>& p, vector<int>& q) {
            return p[0] * p[0] + p[1] * p[1] > q[0] * q[0] + q[1] * q[1];
        }
    };
};

File: 0973-k-closest-points-to-origin/0973-k-closest-points-to-origin.java
// //BRUTE FORCE
// class Solution {
// public int[][] kClosest(int[][] points, int K) {
//     Arrays.sort(points, (p1, p2) -> p1[0] * p1[0] + p1[1] * p1[1] - p2[0] * p2[0] - p2[1] * p2[1]);
//     return Arrays.copyOfRange(points, 0, K);
// }
// }
//MAX HEAP
// class Solution{
// public int[][] kClosest(int[][] points, int K) {
//     PriorityQueue<int[]> pq = new PriorityQueue<int[]>((p1, p2) -> p2[0] * p2[0] + p2[1] * p2[1] - p1[0] * p1[0] - p1[1] * p1[1]);
//     for (int[] p : points) {
//         pq.offer(p);
//         if (pq.size() > K) {
//             pq.poll();
//         }
//     }
//     int[][] res = new int[K][2];
//     while (K > 0) {
//         res[--K] = pq.poll();
//     }
//     return res;
// }

// }

//QUICK SELECT
class Solution {

    public int[][] kClosest(int[][] points, int k) {

        quickSelect(points, 0, points.length - 1, k);

        int[][] result = new int[k][2];

        for (int i = 0; i < k; i++) {
            result[i] = points[i];
        }

        return result;
    }

    private void quickSelect(int[][] points, int left, int right, int k) {

        while (left <= right) {

            int pivotIndex = partition(points, left, right);

            if (pivotIndex == k - 1) {
                return;
            }

            if (pivotIndex > k - 1) {
                right = pivotIndex - 1;
            } else {
                left = pivotIndex + 1;
            }
        }
    }

    private int partition(int[][] points, int left, int right) {

        long pivotDistance =
            (long) points[right][0] * points[right][0] +
            (long) points[right][1] * points[right][1];

        int smaller = left;

        for (int i = left; i < right; i++) {

            long currentDistance =
                (long) points[i][0] * points[i][0] +
                (long) points[i][1] * points[i][1];

            if (currentDistance <= pivotDistance) {
                swap(points, smaller, i);
                smaller++;
            }
        }

        swap(points, smaller, right);

        return smaller;
    }

    private void swap(int[][] points, int i, int j) {

        int[] temp = points[i];
        points[i] = points[j];
        points[j] = temp;
    }
}